controllers/home_test.go

```diff
@@ -6,18 +6,58 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/gutakk/go-google-scraper/db"
+	"github.com/gutakk/go-google-scraper/models"
 	"github.com/gutakk/go-google-scraper/tests"
 	"gopkg.in/go-playground/assert.v1"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
 )
 
-func TestDisplayHome(t *testing.T) {
+func TestDisplayHomeWithoutUserSession(t *testing.T) {
 	engine := tests.GetRouter(true)
 	new(HomeController).applyRoutes(engine)
 
 	w := tests.PerformRequest(engine, "GET", "/", nil, nil)
 	p, err := ioutil.ReadAll(w.Body)
-	pageOK := err == nil && strings.Index(string(p), "<title>Home</title>") > 0
+	isHomePage := err == nil && strings.Index(string(p), "<title>Home</title>") > 0
+	isDisplayGreeting := err == nil && strings.Index(string(p), "<h2 class=\"greeting\">") > 0
+	isDisplayLogin := err == nil && strings.Index(string(p), "<a href=\"/login\">Login</a>") > 0
+	isDisplayRegister := err == nil && strings.Index(string(p), "<a href=\"/register\">Register</a>") > 0
 
 	assert.Equal(t, http.StatusOK, w.Code)
-	assert.Equal(t, true, pageOK)
+	assert.Equal(t, true, isHomePage)
+	assert.Equal(t, true, isDisplayLogin)
+	assert.Equal(t, true, isDisplayRegister)
+	assert.Equal(t, false, isDisplayGreeting)
+}
+
+func TestDisplayHomeWithUserSession(t *testing.T) {
+	testDB, _ := gorm.Open(postgres.Open(tests.ConstructTestDsn()), &gorm.Config{})
+	db.GetDB = func() *gorm.DB {
+		return testDB
+	}
+
+	_ = db.GetDB().AutoMigrate(&models.User{})
+
+	engine := tests.GetRouter(true)
+	new(HomeController).applyRoutes(engine)
+
+	// Cookie from login API Set-Cookie header
+	headers := http.Header{}
+	cookie := "mysession=MTYwNjI3ODk0NHxEdi1CQkFFQ180SUFBUkFCRUFBQUlmLUNBQUVHYzNSeWFXNW5EQWtBQjNWelpYSmZhV1FFZFdsdWRBWUVBUDRCR0E9PXxa_dKXde8j6m4z_kPgaiPYuDGHj79HxhCMNw3zIoeM6g=="
+	headers.Set("Cookie", cookie)
+
+	response := tests.PerformRequest(engine, "GET", "/", headers, nil)
+	p, err := ioutil.ReadAll(response.Body)
+	isHomePage := err == nil && strings.Index(string(p), "<title>Home</title>") > 0
+	isDisplayGreeting := err == nil && strings.Index(string(p), "<h2 class=\"greeting\">") > 0
+	isDisplayLogin := err == nil && strings.Index(string(p), "<a href=\"/login\">Login</a>") > 0
```

Comment:
I am not sure if you plan to test for everything on each page moving forward. That seems more a good case for UI tests while unit tests for controllers would be more limited with just checking for HTTP codes and success/error cases.

Reply:
Fixed in bb2c084d74eec9afbd228621a2289acde50b3515